src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -35,20 +35,42 @@ def compile(request_data)
       def create_node(request_data)
         facts, trusted_facts = process_facts(request_data)
         certname = request_data['certname']
-        environment = request_data['environment'] || 'production'
+        requested_environment = request_data['environment'] || 'production'
         transaction_uuid = request_data['transaction_uuid']
         prefer_requested_environment =
           request_data.dig('options', 'prefer_requested_environment')
 
+
         node = Puppet.override(trusted_information: trusted_facts) do
           Puppet::Node.indirection.find(certname,
-                                        environment: environment,
+                                        environment: requested_environment,
                                         facts: facts,
                                         transaction_uuid: transaction_uuid)
         end
 
+        if request_data['facts'].nil? && !prefer_requested_environment
+          tries = 0
+          environment = requested_environment
+          while node.environment != environment
+            if tries > 3
+              raise Puppet::Error, _("Node environment didn't stabilize after %{tries} fetches, aborting run") % { tries: tries }
+            end
+
+            environment = node.environment
+            facts = get_facts_from_pdb(certname, environment.to_s)
+            node = Puppet.override(trusted_information: trusted_facts) do
```

Comment:
So this is assuming that `trusted_facts` will never change from what was either passed in with the request or pulled out of PDB intitially. I guess that makes sense, considering what they usually consist of?

Reply:
I think the real reason I did this is that I didn't correctly follow the logic of how getting trusted facts can depend on the facts returned from puppetdb. But it does seem true that they shouldn't change.

Reply:
I'm fine leaving this as is then.